src/scf/newton.jl

```diff
@@ -0,0 +1,317 @@
+    unpack(x) = unpack_ψ(basis, x)
+
+    # project rhs on the tangent space before starting
+    proj_tangent!(rhs, ψ)
+    rhs_pack = pack(rhs)
+
+    # preconditioner
+    Pks = [PreconditionerTPA(basis, kpt) for kpt in basis.kpoints]
+    for ik = 1:length(Pks)
+        precondprep!(Pks[ik], ψ[ik])
+    end
+    function f_ldiv!(x, y)
+        δψ = unpack(y)
+        proj_tangent!(δψ, ψ)
+        Pδψ = [ Pks[ik] \ δψk for (ik, δψk) in enumerate(δψ)]
+        proj_tangent!(Pδψ, ψ)
+        x .= pack(Pδψ)
+    end
+
+    # Rayleigh-coefficients
+    Λ = map(enumerate(ψ)) do (ik, ψk)
+        Hk = H.blocks[ik]
+        Hψk = Hk * ψk
+        ψk'Hψk
+    end
+
+    # mapping of the linear system on the tangent space
+    function ΩpK(x)
+        δψ = unpack(x)
+        Kδψ = apply_K(basis, δψ, ψ, ρ, occupation)
+        Ωδψ = apply_Ω(δψ, ψ, H, Λ)
+        pack(Ωδψ + Kδψ)
+    end
+    J = LinearMap{T}(ΩpK, size(rhs_pack, 1))
+
+    # solve (Ω+K) δψ = rhs on the tangent space with CG
+    δψ = cg(J, rhs_pack, Pl=FunctionPreconditioner(f_ldiv!),
+            reltol=tol_cg/norm(rhs), verbose=verbose)
+
+    unpack(δψ)
+end
+
+"""
+    newton(basis::PlaneWaveBasis{T}; ψ0=nothing,
+           tol=1e-6, tol_cg=1e-10, maxiter=20, verbose=false,
+           callback=NewtonDefaultCallback(),
+           is_converged=NewtonConvergenceDensity(tol))
+
+Newton algorithm. Be careful that the starting point needs to be not too far
+from the solution.
+"""
+function newton(basis::PlaneWaveBasis{T}, ψ0;
+                tol=1e-6, tol_cg=1e-10, maxiter=20, verbose=false,
+                callback=NewtonDefaultCallback(),
+                is_converged=NewtonConvergenceDensity(tol)) where T
+
+    # setting parameters
+    model = basis.model
+    @assert model.spin_polarization in (:none, :spinless, :collinear)
+    @assert model.temperature == 0 # temperature is not yet supported
+    filled_occ = filled_occupation(model)
+    n_spin = basis.model.n_spin_components
+    n_bands = div(div(model.n_electrons, filled_occ), n_spin)
+
+    # number of kpoints and occupation
+    Nk = length(basis.kpoints)
+    occupation = [filled_occ * ones(T, n_bands) for ik = 1:Nk]
+
+    # check that there are no virtual orbitals
+    @assert n_bands == size(ψ0[1],2)
+
+    # iterators
+    err = Inf
+    n_iter = 0
+
+    # orbitals, densities and energies to be updated along the iterations
+    ψ = deepcopy(ψ0)
+    ρ = compute_density(basis, ψ, occupation)
+    energies, H = energy_hamiltonian(basis, ψ, occupation; ρ=ρ)
+    converged = false
+
+    # perform iterations
+    while !converged && n_iter < maxiter
+        n_iter += 1
+
+        # compute next step
+        res = compute_projected_gradient(basis, ψ, occupation)
+        # solve (Ω+K) δψ = -res so that the Newton step is ψ <- ψ+δψ
+        δψ = solve_ΩplusK(basis, ψ, -res, occupation; tol_cg=tol_cg,
+                          verbose=verbose)
+        ψ = map(1:Nk) do ik ortho_qr(ψ[ik] + δψ[ik]) end
```

Comment:
I don't like that. In that case better use a list comprehension

Reply:
I'm the one who suggested the map do pattern to Gaspard, so to clarify: the map/do pattern is good when you have a block of code that takes several lines. For one-liners like these, better to use a comprehension